parlai/core/metrics.py

```diff
@@ -530,6 +530,72 @@ def compute(guess: str, answers: List[str]) -> F1Metric:
         return F1Metric(max(f1 for p, r, f1 in scores), 1)
 
 
+class RareWordF1Metric:
+    """
+    Helper class for computing F1 with an emphasis on infrequent words.
+    """
+
+    def __init__(self, corpus: str, top_p: Optional[float] = 0.5):
+        try:
+            import nltk
+        except ImportError:
+            raise ImportError('Please install nltk (e.g. pip install nltk).')
+        words = normalize_answer(corpus).split()
+        self._freq_dist = nltk.FreqDist(words)
+        self._cutoff = RareWordF1Metric._find_cutoff_count(self._freq_dist, top_p)
+
+    @staticmethod
+    def _find_cutoff_count(freq_dist, top_p: float) -> int:
+        """
+        Finds the word occurance for which the cumulative occurances
+        are `top_p` of the overall word count.
+        """
+        assert top_p < 1
+        target = sum(freq_dist.values()) * top_p
+        cumul = 0
+        for _, v in freq_dist.most_common():
+            cumul += v
+            if cumul > target:
+                return v
+
+    @staticmethod
+    def _rarity_weight(freq_dist, word: str, cutoff: int) -> float:
+        """
+        A score multiplier that signifies how rare a word is.
+        The words with more than `cutoff` occurances in the corpus will
+        have a weight of 0, and very rare words will have a weight near 1.
+        """
+        return max(0, (cutoff - freq_dist[word]) / cutoff)
+
+    def _weighted_f1_score(self, pred_items: List[str], gold_items: List[str]) -> float:
+        """
+        Compute f1 given a set of gold and prediction items, weighted by the infrequency of each word.
+        """
+        weights = {
+            w: RareWordF1Metric._rarity_weight(
+                freq_dist=self._freq_dist, word=w, cutoff=self._cutoff
+            )
+            for w in set(pred_items + gold_items)
+        }
+        common = Counter(gold_items) & Counter(pred_items)
+        weighted_common = {w: c * weights[w] for w, c in common.items()}
+        true_pos_score = sum(weighted_common.values())
+        if true_pos_score == 0:
+            return 0
+        precision = true_pos_score / sum(weights[w] for w in pred_items)
```

Comment:
oh, this is really interesting, i had originally imagined weights to just be 1 if above the threshold, 0 otherwise, i wonder if you could make that a configurable option for the metric? this is a cool way of doing it too

Reply:
Glad you like it! If you look at my plots of `_rarity_weight` in the PR description, you'll see that it rises fairly sharply after the chosen threshold. I was afraid that this might tend toward the simpler version (0 above and 1 below the threshold) in most cases, at which point the weighting would complicate the metric without adding much value. So I just opted for the simpler version for now.

But if there were a way of calculating the rarity weight which rose to 1.0 more gradually, we could try another version of this metric that doesn't have a "cutoff" and only has the weighting. That would feel much more elegant and would probably be more robust to different word distributions.

The issue I was running into in trying to find a function like that, though, was that the top few words of the distribution are _so_ common relative to even the rest of the top 50 words that I have to find a function that pushes the majority of the range (let's say 0.0 to 0.99) down to near zero while expanding the last bit of the range (0.99 to 1.0) to a gradual slope from near-zero to 1.0. Otherwise fairly common words are still given a high weight.